File: GateSyntaxWebb/src/runtime/state_store.ts
// Reactive state store — mirrors GateSyntax.Runtime.StateStore.cs

type Listener = (value: unknown) => void;

export class StateStore {
  private data      = new Map<string, unknown>();
  private listeners = new Map<string, Set<Listener>>();

  set(name: string, value: unknown): void {
    const key = name.toUpperCase();
    if (this.data.get(key) === value) return;
    this.data.set(key, value);
    this.listeners.get(key)?.forEach(fn => fn(value));
  }

  get(name: string, defaultValue: unknown = ''): unknown {
    return this.data.has(name.toUpperCase())
      ? this.data.get(name.toUpperCase())
      : defaultValue;
  }

  setDefault(name: string, value: unknown): void {
    const key = name.toUpperCase();
    if (!this.data.has(key)) this.data.set(key, value);
  }

  /** Returns an unsubscribe function. */
  subscribe(name: string, fn: Listener): () => void {
    const key = name.toUpperCase();
    if (!this.listeners.has(key)) this.listeners.set(key, new Set());
    this.listeners.get(key)!.add(fn);
    return () => this.listeners.get(key)?.delete(fn);
  }

  snapshot(): Record<string, unknown> {
    return Object.fromEntries(this.data);
  }

  restore(saved: Record<string, unknown>): void {
    for (const [k, v] of Object.entries(saved)) {
      this.data.set(k.toUpperCase(), v);
    }
  }
}

File: GateSyntaxJava/core/com/gatesyntax/runtime/StateStore.java
package com.gatesyntax.runtime;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

/** Reactive state store — mirrors GateSyntax.Runtime.StateStore. */
public class StateStore {

    private final Map<String, Object>            data      = new ConcurrentHashMap<>();
    private final Map<String, List<Consumer<Object>>> subs = new ConcurrentHashMap<>();

    public void set(String name, Object value) {
        String key = name.toUpperCase();
        Object old = data.get(key);
        if (Objects.equals(old, value)) return;
        data.put(key, value);
        List<Consumer<Object>> listeners = subs.get(key);
        if (listeners != null) listeners.forEach(fn -> fn.accept(value));
    }

    public Object get(String name) {
        return data.getOrDefault(name.toUpperCase(), "");
    }

    public Object get(String name, Object defaultValue) {
        return data.getOrDefault(name.toUpperCase(), defaultValue);
    }

    public void setDefault(String name, Object value) {
        data.putIfAbsent(name.toUpperCase(), value);
    }

    public void subscribe(String name, Consumer<Object> listener) {
        subs.computeIfAbsent(name.toUpperCase(), k -> new CopyOnWriteArrayList<>()).add(listener);
    }

    public Map<String, Object> snapshot() {
        return Collections.unmodifiableMap(new HashMap<>(data));
    }
}
